File: Python/pattern-matching.py
# Naive Search

def naive_search(y, x, n, m):
    i = 0
    while i <= n - m:
        j = 0
        while j < m and x[j] == y[j + i]:
            j += 1
        if j == m:
                print 'y occurs in x at the position %d' % (i)
        i += 1

# Compute borders of a string

def borders(x, m):
    i = 0
    j = 0
    border = []
    for j in range (1, m):
        border.append(i)
        while i >= 0 and x[j] != x[i]:
            if i == 0:
                i = -1
            else:
                i = border[i - 1]
        i = i + 1
    border.append(i)
    return border

# Preprocessing Morris-Pratt Algorithms

def mp_borders(x, m):
    j = -1
    i = 0
    mpNext = [j]
    for i in range(0, m - 1):
        while j >= 0 and x[i] != x[j]:
            j = mpNext[j]
        j += 1
        mpNext.append(j)
    mpNext.append(j + 1)
    return mpNext


# Preprocessing Knuth-Morris-Pratt Algorithm

def kmp_borders(x, m):
    i = 0
    j = -1
    kmpNext = [j]
    for i in range(0, m - 1):
        while j >= 0 and x[i] != x[j]:
            j = kmpNext[j]
        i += 1
        j += 1
        if x[i] == x[j]:
            kmpNext.append(kmpNext[j])
        else:
            kmpNext.append(j)
    return kmpNext


# KMP and MP Algorithms Search function

def search(x, y, m, n, algorithm):

    print "Preprocessing phase:"
    if algorithm == 'kmp':
        MP_next = kmp_borders(x, m)
    elif algorithm == 'mp':
        MP_next = mp_borders(x, m)
    print MP_next

    print "\nSearching phase:"
    i = 0
    j = 0
    while j < m:
        while (i == n) or (i >= 0 and x[j] != y[i]):
            i = MP_next[i]
        i = i + 1
        j = j + 1
        if i == n:
            print "x occurs in y at the position %d" % (j - 1)


# Table of prefixes

def prefixes(x, m):
    pref = [m]
    g = 0
    f = 0
    i = 1
    while i < m:
        if i < g and pref[i - f] != g - i:
            pref.append(min(pref[i - f], g - i))
        else:
            g = max(g, i)
            f = i
            while g < m and x[g] == x[g - f]:
                g = g + 1
            pref.append(g - f)
        i = i + 1
    return pref

File: Javascript/pattern-matching.js
/*
 **  Naive Search
 */

function NaiveSearch(x, y, n, m) {
    var i = 0;
    while (i <= n - m) {
        var j = 0;
        while ((j < m) && (y[j] == x[i + j])) {
            j++;
        }
        if (j == m) {
            console.log('y occurs in x at the position ' + i);
        }
        i++;
    }
}


/*
 ** Compute borders of a string
 */

function Borders(x, m) {
    var i, j = 1;
    var border = [];
    while (j < m) {
        border.push(i);
        while (i >= 0 && x[j] !== x[i]) {
            if (!i) {
                i--;
            } else {
                i = border[i - 1];
            }
        }
        i++;
        j++;
    }
    border.push(i);
    return border;
}

/*
 ** Preprocessing Morris-Pratt Algorithm
 */

function MP_Next(x, m) {
    var j = -1,
        i = 0;
    var mpNext = [j];
    while (i < m - 1) {
        while (j >= 0 && x[i] !== x[j]) {
            j = mpNext[j];
        }
        j++;
        i++;
        mpNext.push(j);
    }
    mpNext.push(j);
    return mpNext;
}

/*
 ** Preprocessing Knuth-Morris-Pratt Algorithm
 */

function KMP_Next(x, m) {
    var j = -1,
        i = 0;
    var kmpNext = [j];
    while (i < m - 1) {
        while (j >= 0 && x[i] !== x[j]) {
            j = mpNext[j];
        }
        j++;
        i++;
        if (x[i] == x[j]) {
            kmpNext.push(kmpNext[j]);
        } else {
            kmpNext.push(j);
        }
        return kmpNext;
    }
}

/*
 ** KMP and MP Algorithms Search function
 */


function Search(x, y, m, n, algorithm) {

    var borders;

    if (algorithm == 'kmp') {
        borders = KMP_Next(x, m);
    } else if (algorithm == 'mp') {
        borders = MP_Next(x, m);
    }

    var i = 0,
        j = 0;

    while (j < m) {
        while (i == n || (i >= 0 && x[j] != y[i])) {
            i = borders[i];
        }

        i++;
        j++;

        if (i == n) {
            console.log("x occurs in y at the position " + (j - 1));
        }
    }
}